File: src/physics/forcing.ts
import { JointForcing } from '../types';

// Deterministic part of a joint torque at absolute time t.
// tau_i(t) = A_i sin(2*pi*f_i*t + phi_i) for the sinusoidal case.
export function deterministicTorque(cfg: JointForcing, t: number): number {
  switch (cfg.type) {
    case 'constant':
      return cfg.amplitude;
    case 'sine':
      return cfg.amplitude * Math.sin(2 * Math.PI * cfg.frequency * t + cfg.phase);
    case 'square': {
      const s = Math.sin(2 * Math.PI * cfg.frequency * t + cfg.phase);
      return cfg.amplitude * (s >= 0 ? 1 : -1);
    }
    case 'none':
    case 'random':
    case 'impulse':
    default:
      return 0;
  }
}

// Random torque given a seeded uniform sample in [0,1). Held constant over an
// integration step for reproducibility. Maps to symmetric noise in [-I, I].
export function randomTorque(cfg: JointForcing, sample: number): number {
  return cfg.type === 'random' ? cfg.noiseIntensity * (sample * 2 - 1) : 0;
}

File: python/double_pendulum/forcing.py
"""Joint torque forcing terms (deterministic + seeded random)."""

from __future__ import annotations

import math

from .types import JointForcing


def deterministic_torque(cfg: JointForcing, t: float) -> float:
    """Deterministic part of a joint torque at absolute time ``t``.

    For the sinusoidal case ``tau_i(t) = A_i sin(2*pi*f_i*t + phi_i)``.
    """

    if cfg.type == "constant":
        return cfg.amplitude
    if cfg.type == "sine":
        return cfg.amplitude * math.sin(2 * math.pi * cfg.frequency * t + cfg.phase)
    if cfg.type == "square":
        s = math.sin(2 * math.pi * cfg.frequency * t + cfg.phase)
        return cfg.amplitude * (1.0 if s >= 0 else -1.0)
    # "none", "random", "impulse" contribute no deterministic term here.
    return 0.0


def random_torque(cfg: JointForcing, sample: float) -> float:
    """Random torque from a seeded uniform sample in ``[0, 1)``.

    Held constant over an integration step for reproducibility. Maps to
    symmetric noise in ``[-I, I]``.
    """

    return cfg.noise_intensity * (sample * 2 - 1) if cfg.type == "random" else 0.0


__all__ = ["deterministic_torque", "random_torque"]
